lib/srv/app/azure/handler.go

```diff
@@ -0,0 +1,298 @@
+	"github.com/gravitational/teleport/lib/utils"
+	awsutils "github.com/gravitational/teleport/lib/utils/aws"
+)
+
+// ForwarderConfig is the Forwarder configuration.
+type ForwarderConfig struct {
+	// Client is an HTTP client instance used for HTTP calls.
+	Client *http.Client
+	// Log is the Logger.
+	Log logrus.FieldLogger
+	// Clock is used to override time in tests.
+	Clock clockwork.Clock
+
+	// getAccessToken is a function for getting access token, pluggable for the sake of testing.
+	getAccessToken getAccessTokenFunc
+}
+
+// CheckAndSetDefaults validates the ForwarderConfig config.
+func (s *ForwarderConfig) CheckAndSetDefaults() error {
+	if s.Client == nil {
+		tr, err := defaults.Transport()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		s.Client = &http.Client{
+			Transport: tr,
+		}
+	}
+	if s.Clock == nil {
+		s.Clock = clockwork.NewRealClock()
+	}
+	if s.Log == nil {
+		s.Log = logrus.WithField(trace.Component, "azure:fwd")
+	}
+	if s.getAccessToken == nil {
+		s.getAccessToken = getAccessTokenManagedIdentity
+	}
+	return nil
+}
+
+// Forwarder is an AWS CLI proxy service that signs AWS requests
+// based on user identity.
+type Forwarder struct {
+	// ForwarderConfig is the Forwarder configuration.
+	ForwarderConfig
+
+	// Forwarder signs and forwards the request to Azure API.
+	*forward.Forwarder
+
+	// tokenCache caches access tokens.
+	tokenCache *utils.FnCache
+}
+
+// NewForwarder creates a new instance of Forwarder.
+func NewForwarder(ctx context.Context, config ForwarderConfig) (*Forwarder, error) {
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	tokenCache, err := utils.NewFnCache(utils.FnCacheConfig{
+		TTL:     time.Second * 60,
+		Clock:   config.Clock,
+		Context: ctx,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	svc := &Forwarder{
+		ForwarderConfig: config,
+		tokenCache:      tokenCache,
+	}
+
+	fwd, err := forward.New(
+		forward.RoundTripper(svc),
+		forward.ErrorHandler(oxyutils.ErrorHandlerFunc(svc.formatForwardResponseError)),
+		forward.PassHostHeader(true),
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	svc.Forwarder = fwd
+	return svc, nil
+}
+
+// RoundTrip handles incoming requests and forwards them to the proper API.
+func (s *Forwarder) RoundTrip(req *http.Request) (*http.Response, error) {
+	sessionCtx, err := common.GetSessionContext(req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	fwdRequest, err := s.prepareForwardRequest(req, sessionCtx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	resp, err := s.Client.Do(fwdRequest)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := sessionCtx.Audit.OnRequest(req.Context(), sessionCtx, fwdRequest, resp, nil); err != nil {
+		s.Log.WithError(err).Warn("Failed to emit audit event.")
+	}
+
+	return resp, nil
+}
+
+func (s *Forwarder) formatForwardResponseError(rw http.ResponseWriter, r *http.Request, err error) {
```

Comment:
Yeah, we will have an analogous problem here with Azure (and GCP next).

I wonder if we can attach this error message to a custom response header (`x-teleport-api-error`?) and sniff that header on `tsh` side?

Reply:
I think that's a great idea.

Reply:
My first attempt: https://github.com/gravitational/teleport/pull/18675/commits/8af456faf46f0d049f41dba37111c6bfb4a4a5c9

```
$ tsh proxy azure --port 3500
Started Azure proxy on http://127.0.0.1:3500.

Use the following credentials and HTTPS proxy setting to connect to the proxy:

  export AZURE_CONFIG_DIR=/Users/tener/.azure-teleport-azure
  export HTTPS_PROXY=http://127.0.0.1:3500
  export MSI_ENDPOINT=https://azure-msi.teleport.dev/my-own-secret
  export REQUESTS_CA_BUNDLE=/Users/tener/.tsh/keys/boson.tener.io/tener-app/boson.tener.io/azure-localca.pem

[2022/12/06 13:47:58] local_proxy.go:269 [Warn] ERROR REPORT: Original Error: context.deadlineExceededError context deadline exceeded Stack Trace:
	github.com/gravitational/teleport/lib/srv/app/azure/handler.go:249 github.com/gravitational/teleport/lib/srv/app/azure.(*Forwarder).replaceAuthHeaders
	github.com/gravitational/teleport/lib/srv/app/azure/handler.go:205 github.com/gravitational/teleport/lib/srv/app/azure.(*Forwarder).prepareForwardRequest
	github.com/gravitational/teleport/lib/srv/app/azure/handler.go:151 github.com/gravitational/teleport/lib/srv/app/azure.(*Forwarder).RoundTrip
	github.com/gravitational/oxy@v0.0.0-20221029012416-9fbf4c444680/forward/fwd.go:174 github.com/gravitational/oxy/forward.(*httpForwarder).serveHTTP
	github.com/gravitational/oxy@v0.0.0-20221029012416-9fbf4c444680/forward/fwd.go:167 github.com/gravitational/oxy/forward.(*Forwarder).ServeHTTP
	github.com/gravitational/teleport/lib/srv/app/server.go:853 github.com/gravitational/teleport/lib/srv/app.(*Server).serveSession
	github.com/gravitational/teleport/lib/srv/app/server.go:815 github.com/gravitational/teleport/lib/srv/app.(*Server).serveHTTP
	github.com/gravitational/teleport/lib/srv/app/server.go:778 github.com/gravitational/teleport/lib/srv/app.(*Server).ServeHTTP
	github.com/gravitational/teleport/lib/auth/middleware.go:617 github.com/gravitational/teleport/lib/auth.(*Middleware).ServeHTTP
	github.com/gravitational/teleport/lib/httplib/httplib.go:93 github.com/gravitational/teleport/lib/httplib.MakeTracingHandler.func1
	net/http/server.go:2109 net/http.HandlerFunc.ServeHTTP
	go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.36.4/handler.go:204 go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.(*Handler).ServeHTTP
	net/http/server.go:2947 net/http.serverHandler.ServeHTTP
	net/http/server.go:1991 net/http.(*conn).serve
	runtime/asm_arm64.s:1172 runtime.goexit

	User Message: context deadline exceeded
```

I'm slightly worried this may be too much information for the end user, though. What do you think?

Reply:
Thanks a lot for addressing this!

> I'm slightly worried this may be too much information for the end user, though.

Personally, I think just the error message is good enough. The client-side user cannot do much with the stack trace of the server anyway. That said I don't mind having it either.

Either way, I think we should be clear that this error comes from the server's response.

Reply:
I'm having mixed feelings about exposing stack and internal teleport error context to the tsh user. 
Most likely it won't bring any value to the user because `ERROR REPORT: Original Error: context.deadlineExceededError context deadline exceeded Stack Trace:` doesn't mean anything to the `tsh az` user that doesn't know Teleport/GO context. Also, there is a risk that the debug log will leak some kind of internal information about teleport or Azure. Though I agree that  `x-teleport-api-error` is a mechanism to return the error to the user but we should have more control over what we are returning and be more verbose in error message.

Reply:
@smallinsky 

I was following https://github.com/gravitational/teleport/issues/16571 here, where it notes that the fallback to diagnosing the errors is checking the service logs:

> What would you like Teleport to do? Improve the AWS CLI proxy (`tsh aws`)'s error logging. At the moment, when a command fails we aren't getting anything useful back, instead we get `error 500 when calling XYZ operation`. We are still able to find the real error by checking `Teleport` service logs on the app proxy node which shows the actual error: `[AWS:SIGNE] Failed to process request: Post "https://rds.us-west-1c.amazonaws.com/": dial tcp: lookup [rds.us-west-1c.amazonaws.com](http://rds.us-west-1c.amazonaws.com/) on 17x.10.0.2:5x: no such host`.

What is your proposal to address https://github.com/gravitational/teleport/issues/16571? Return only the `err.Error()` string? Something else?

Reply:
I think that we should implement more suited mechanism for error propagation instead of forwarding the server error to the aws/azure cli client. 

For instance we can use a user error and in case of server error create a descriptive user CLI error on the server and forward it using `x-teleport-api-error` mechanism. But after a second thought we can start with current solution and alter this approach.